### ISSUE TITLE
vllm_golden_gate: marginal near-tie position intermittently violates the confident bound

### ISSUE THREAD

**Issue by xiaguan** [xiaguan/pegainfer#286]
## Problem

`pegainfer-kimi-k2/tests/vllm_golden_gate.rs` teacher-forced sweep intermittently fails on a near-tie position. Observed on 8×H200 (K2.6 TP1/DP8 PPLX), 4 runs across two adjacent commits with no forward-path diff between them:

| Run | Commit | Result |
|---|---|---|
| 1 | sampling branch | **FAIL** — `long-en` pos 27: picked 1364, regret 0.7500 > bound 0.4435 (top-1 lp -0.41) |
| 2 | main (59fbbb5) | pass (6 in-bound flips) |
| 3 | sampling branch | pass (6 in-bound flips) |
| 4 | sampling branch | pass (9 in-bound flips) |

The teacher-forced |Δlogprob| tail stats are identical to 4 decimals across all runs (p99 0.2878, max 0.5150, flip-pick max 1.5743) — the logit field is stable; only near-tie argmax picks wobble run-to-run (flip count 6→9), most likely atomic-reduction order (Marlin split-K / PPLX combine). At `long-en` pos 27 pegainfer has tokens 316/1364 within jitter distance while vLLM scores them 0.75 nat apart, so the unlucky pick lands above the confident-position bound.

## Repro

`PEGAINFER_TEST_MODEL_PATH=<K2.6> cargo test --release -p pegainfer-kimi-k2 --features kimi-k2 --test vllm_golden_gate` — repeat; failure rate ~1/3 in this sample.

## Acceptance

- The gate distinguishes "pegainfer's pick is near-tied with vLLM's argmax under pegainfer's own logits" from a real regression — e.g. exempt flips where pegainfer's top-2 gap < ε, or score the flip under pegainfer's logits as a second signal.
- 5 consecutive green runs on the same binary on 8×H200.

**Comment by xiaguan** [comment]
New evidence reframes this issue: it is not a marginal near-tie — it is an intermittent **final-position corruption in the concurrent pass**, present on `main` (static KV arena) and on `feat/kimi-kv-pool` (paged) alike.

6-run A/B on one 8×H200 node (3× `c70734a` = pre-paged-KV main, 3× branch `a688200`):

| run | arena | violations |
|---|---|---|
| main r1 | static | translation@31: picked 6, 7.375 nat |
| main r2 | static | translation@31: 11214, 1.375 + long-en@31: 1481, 4.25 |
| main r3 | static | long-zh@20 (teacher-forced): 0.625 + translation@31: 11214, 1.375 |
| branch r2 | paged | long-zh@20: 0.625 + json@5: 1.25 + translation@31: 6, 7.375 + long-en@31: 382, 8.375 |
| branch r3 | paged | translation@31: 6, 7.375 |
| branch r4 | paged | translation@31: 11214, 1.375 |

Observations:

- translation pos 31 violates in **6/6 runs on both arenas**, with varying magnitude (1.4–7.4 nat) and varying picked token (6 / 11214 / 382 / 1481). Reaching pos 31 requires exact token match through pos 30, so the contexts were identical — the engine's own argmax at the final position is sometimes a token vLLM scores 7+ nat down. That is a corrupted distribution, not ULP wobble.
- Only the concurrent pass is affected (bs=1 parity: 0 violations in all 6 runs). Teacher-forced long-zh@20 at 0.625 vs bound 0.5465 straddles on both arenas — that part is the marginal class originally described here.
- Node-correlated: the same gate was green 2/2 on a different 8×H200 node on 2026-06-05 (#269). On this node it is red 7/7 across both arenas.

Implications:

- The A/B protocol in this issue worked as designed: the paged-KV branch (PR #292) shows the same profile as main and is not the cause.
- The final-position concurrent corruption needs its own investigation: likely interactions when co-batched requests finish (batch recomposition at the last decode step). Severity varies run-to-run, so single-run green is weak evidence of absence.

**Comment by xiaguan** [comment]
Additional A/B evidence (8×H200, 2026-06-07, during #230 verification):

- Branch tip (prefix caching #230, 3 runs): concurrent-pass violations at **translation pos 31** in 3/3 (picks 6/11214/6, 1.4–7.4 nat), plus **json pos 5** (pick 9049, 1.25 nat) in 1/3.
- Pre-#230 commit 96143e6 (paged pool, no prefix cache, 2 runs): r1 fully green; r2 red with the **identical profile** — json pos 5 pick 9049 AND translation pos 31 pick 11214.
- Teacher-forced sweep and bs=1 parity: 0 violations in all 5 runs.

json pos 5 is therefore part of the same pre-existing intermittent concurrent corruption, not a #230/#292 regression — same positions, same picked tokens, fires with and without prefix caching. Frequency on this node today: 4/5 red across both arms (the earlier 6-run A/B was 6/6 on both arenas), still node/day-correlated and confined to the concurrent pass.

**Comment by xiaguan** [comment]
**Reproduced 2026-06-08 (8×H200, DeepEP + graph checkout).** `vllm_golden_gate` re-run:
- teacher-forced sweep: **0 violations** (384 pos, 375 exact = 97.7%, |Δlogprob| mean 0.032 / p99 0.324).
- bs=1 greedy parity: **0 violations**.
- concurrent free-greedy: **1 violation at `translation` pos 31** — pegainfer picked 11214, which vLLM scores 1.375 nat below its argmax 30949.

Same final-position, concurrent-only signature on the DeepEP backend — consistent with the migration record (teacher-forced + bs1 clean, concurrent red at the same positions). Confirms a concurrent-decode final-position mispick, not gate flakiness or a backend regression. The title undersells it: the cheap half is a near-tie exemption in `check_pick()`, but the real bug is the concurrent-batch corruption. Tracked in the refreshed roadmap (#306).
